File: src/main/java/antlr/TreepatVisitorImplementation.java
package antlr;

import antlr.generate.TreepatParser;
import antlr.generate.TreepatVisitor;
import ast.ASTNode;
import ast.Child;
import ast.Node;
import ast.Sibling;
import org.antlr.v4.runtime.tree.ErrorNode;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.RuleNode;
import org.antlr.v4.runtime.tree.TerminalNode;
import tree.ImpTargetTreeNode;
import tree.TargetTreeNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static java.util.stream.Collectors.toList;

public class TreepatVisitorImplementation implements TreepatVisitor<ASTNode> {
    @Override
    public ASTNode visitModel(TreepatParser.ModelContext ctx) {
        TargetTreeNode targetTreeNode = new ImpTargetTreeNode();
        ctx.subtree().accept(this).execute(targetTreeNode);
        return null;
    }

    @Override
    public ASTNode visitSubtree(TreepatParser.SubtreeContext ctx) {
        ASTNode expression = ctx.expression().accept(this);
        if( ctx.child() == null )
        {
            return expression;
        }
        ASTNode child = ctx.child().accept(this);
        return new Child(expression, child);
    }

    @Override
    public ASTNode visitExpression(TreepatParser.ExpressionContext ctx) {
        if( ctx.simple_expression() != null )
        {
            return ctx.simple_expression().accept(this);
        }
        return ctx.depth_closure().accept(this);
    }

    @Override
    public ASTNode visitChild(TreepatParser.ChildContext ctx) {
        return ctx.sibling().accept(this);
    }

    @Override
    public ASTNode visitSibling(TreepatParser.SiblingContext ctx) {
        List<ASTNode> siblings;
        siblings = ctx.union().stream()
                .map(instruction -> instruction.accept(this))
                .collect(toList());
        return new Sibling(siblings);
    }

    @Override
    public ASTNode visitUnion(TreepatParser.UnionContext ctx) {
        List<ASTNode> nodes;
        nodes = ctx.subtree_wrapper().stream()
                .map(node -> node.accept(this))
                .collect(toList());
        return nodes.get(0);
    }

    @Override
    public ASTNode visitSubtree_wrapper(TreepatParser.Subtree_wrapperContext ctx) {
        return ctx.subtree().accept(this);
    }

    @Override
    public ASTNode visitDepth_closure(TreepatParser.Depth_closureContext ctx) {
        return ctx.child().accept(this);
    }

    @Override
    public ASTNode visitSimple_expression(TreepatParser.Simple_expressionContext ctx) {
        ASTNode value_to_return = null;
        if( ctx.term() != null )
        {
            value_to_return =ctx.term().accept(this);;
        }
        else if( ctx.breadth_closure() != null )
        {
            value_to_return = ctx.breadth_closure().accept(this);
        }
        else if( ctx.depth_term() != null )
        {
            value_to_return = ctx.depth_term().accept(this);
        }
        return  value_to_return;
    }

    @Override
    public ASTNode visitDepth_term(TreepatParser.Depth_termContext ctx) {
        return ctx.term().accept(this);
    }

    @Override
    public ASTNode visitBreadth_closure(TreepatParser.Breadth_closureContext ctx) {
        return ctx.term().accept(this);
    }

    @Override
    public ASTNode visitTerm(TreepatParser.TermContext ctx) {
        return ctx.node().accept(this);
    }

    @Override
    public ASTNode visitNode(TreepatParser.NodeContext ctx) {
        return new Node(ctx.name.getText());
    }

    @Override
    public ASTNode visit(ParseTree parseTree) {
        return parseTree.accept(this);
    }

    @Override
    public ASTNode visitChildren(RuleNode ruleNode) {
        return null;
    }

    @Override
    public ASTNode visitTerminal(TerminalNode terminalNode) {
        return null;
    }

    @Override
    public ASTNode visitErrorNode(ErrorNode errorNode) {
        return null;
    }
}

File: src/main/java/antlr/TreepatVisitorImplementation.kt
package antlr

import antlr.generate.TreepatParser
import antlr.generate.TreepatVisitor
import ast.ASTNode
import ast.Child
import ast.Node
import ast.Sibling
import java.util.stream.Collectors
import org.antlr.v4.runtime.tree.ErrorNode
import org.antlr.v4.runtime.tree.ParseTree
import org.antlr.v4.runtime.tree.RuleNode
import org.antlr.v4.runtime.tree.TerminalNode

class TreepatVisitorImplementation : TreepatVisitor<ASTNode?> {
    override fun visitModel(ctx: TreepatParser.ModelContext): ASTNode {
        return ctx.subtree().accept<ASTNode>(this)
    }

    override fun visitSubtree(ctx: TreepatParser.SubtreeContext): ASTNode {
        val expression = ctx.expression().accept<ASTNode>(this)
        if (ctx.child() == null) {
            return expression
        }
        val child = ctx.child().accept<ASTNode>(this)
        return Child(expression, child)
    }

    override fun visitExpression(ctx: TreepatParser.ExpressionContext): ASTNode {
        return if (ctx.simpleExpression() != null) {
            ctx.simpleExpression().accept<ASTNode>(this)
        } else ctx.depthClosure().accept<ASTNode>(this)
    }

    override fun visitChild(ctx: TreepatParser.ChildContext): ASTNode {
        return ctx.sibling().accept<ASTNode>(this)
    }

    override fun visitSibling(ctx: TreepatParser.SiblingContext): ASTNode {
        val siblings: List<ASTNode>
        siblings = ctx.union().stream()
            .map { instruction: TreepatParser.UnionContext -> instruction.accept<ASTNode>(this) }
            .collect(Collectors.toList())
        return Sibling(siblings)
    }

    override fun visitUnion(ctx: TreepatParser.UnionContext): ASTNode {
        val nodes: List<ASTNode>
        nodes = ctx.subtreeWrapper().stream()
            .map { node: TreepatParser.SubtreeWrapperContext -> node.accept<ASTNode>(this) }
            .collect(Collectors.toList())
        return nodes[0]
    }

    override fun visitSubtreeWrapper(ctx: TreepatParser.SubtreeWrapperContext): ASTNode {
        return ctx.subtree().accept<ASTNode>(this)
    }

    override fun visitDepthClosure(ctx: TreepatParser.DepthClosureContext): ASTNode {
        return ctx.child().accept<ASTNode>(this)
    }

    override fun visitSimpleExpression(ctx: TreepatParser.SimpleExpressionContext): ASTNode {
        var valueTo_return: ASTNode? = null
        if (ctx.term() != null) {
            valueTo_return = ctx.term().accept<ASTNode>(this)
        } else if (ctx.breadthClosure() != null) {
            valueTo_return = ctx.breadthClosure().accept<ASTNode>(this)
        } else if (ctx.depthTerm() != null) {
            valueTo_return = ctx.depthTerm().accept<ASTNode>(this)
        }
        return valueTo_return!!
    }

    override fun visitDepthTerm(ctx: TreepatParser.DepthTermContext): ASTNode {
        return ctx.term().accept<ASTNode>(this)
    }

    override fun visitBreadthClosure(ctx: TreepatParser.BreadthClosureContext): ASTNode {
        return ctx.term().accept<ASTNode>(this)
    }

    override fun visitTerm(ctx: TreepatParser.TermContext): ASTNode {
        return ctx.node().accept<ASTNode>(this)
    }

    override fun visitNode(ctx: TreepatParser.NodeContext): ASTNode {
        return Node(ctx.name.text)
    }

    override fun visit(parseTree: ParseTree): ASTNode {
        return parseTree.accept<ASTNode>(this)
    }

    override fun visitChildren(ruleNode: RuleNode): ASTNode? {
        return null
    }

    override fun visitTerminal(terminalNode: TerminalNode): ASTNode? {
        return null
    }

    override fun visitErrorNode(errorNode: ErrorNode): ASTNode? {
        return null
    }
}
